wgpu/src/api/buffer.rs

```diff
@@ -640,6 +652,53 @@ impl<'a> From<BufferSlice<'a>> for crate::BindingResource<'a> {
     }
 }
 
+fn range_overlaps(a: &Range<BufferAddress>, b: &Range<BufferAddress>) -> bool {
+    a.start < b.end && b.start < a.end
+}
+
+#[derive(Debug, Copy, Clone)]
+enum RangeMappingKind {
+    Mutable,
+    Immutable,
+}
+
+impl RangeMappingKind {
+    /// Returns true if a range of this kind can touch the same bytes as a range of the other kind.
+    ///
+    /// This is Rust's Mutable XOR Immutable rule.
```

Comment:
Nit: The phrase is "mutable XOR shared".

(And some pedants point out that the actual logical rule is "not mutable OR not shared" because it's always okay to have something not borrowed at all.)